apps/slackbot/utils.py

```diff
@@ -14,6 +22,108 @@ def load_files(files: List[str]) -> List[Document]:
         else:
             raise NotImplementedError(f"File type {f} not supported")
         documents.extend(loader.load())
-    
+
     print(documents)
-    return documents
\ No newline at end of file
+    return documents
+
+
+def get_links_from_string(text):
+    # Define the regular expression pattern to find links inside angle brackets
+    pattern = r'<([^>]*)>'
+
+    # Use re.findall to extract all matches of the pattern in the input string
+    matches = re.findall(pattern, text)
+
+    # Filter the matches to keep only the ones that start with "http://" or "https://"
+    # links = [match for match in matches if match.startswith(
+    #     "http://") or match.startswith("https://")]
+    links = []
+
+    for match in matches:
+        if match.startswith("http://") or match.startswith("https://"):
+            links.append({"url": match, "base_url": get_base_url(match)})
+    return links
+
+
+def get_base_url(link):
+    parsed_url = urlparse(link)
+    base_url = f"{parsed_url.scheme}://{parsed_url.netloc}"
+    return base_url
+
+
+def get_link_from_slack_client_conversation(data):
+    links = []
+    for item in data:
```

Comment:
what is the nested for loop here @Eyobyb can we kind of try optimising this

Reply:
the slack message structure is like that .. no way of joining about it

Reply:
There are almost 4 nested for loops here man

Reply:
[ ##1 loop
  {
    'client_msg_id': '960d41d8-2718-4ce5-b5d5-ff66aa09af24',
    'type': 'message',
    'text': '<@U05HDFV64AU> what does <https://www.reuters.com/world/africa/ethiopia-declares-state-emergency-following-militia-clashes-2023-08-04/> talk about ethiopia',
    'user': 'U05GGHG7T1C',
    'ts': '1691420055.909099',
    'blocks': [
 ##2 loop
      {
        'type': 'rich_text',
        'block_id': 'CCqFe',
        'elements': [
 ##3 loop
          {
            'type': 'rich_text_section',
            'elements': [
 ##4 loop
              {
                'type': 'user',
                'user_id': 'U05HDFV64AU'
              },
              {
                'type': 'text',
                'text': ' what does '
              },
              {
                'type': 'link',
                'url': 'https://www.reuters.com/world/africa/ethiopia-declares-state-emergency-following-militia-clashes-2023-08-04/'
              },
              {
                'type': 'text',
                'text': ' talk about ethiopia'
              }
            ]
          }
        ]
      }
    ],]

Reply:
you can see it from this structure in order to get the link you have to go down for nested arrays

Reply:
Okay let me look into this and try optimizing it then thanks will look into this